pallets/liquidity-pools-forwarder/src/lib.rs

```diff
@@ -0,0 +1,266 @@
+		///
+		/// Origin: Admin.
+		///
+		/// NOTE: Simple weight due to origin requirement.
+		#[pallet::weight(T::DbWeight::get().writes(1))]
+		#[pallet::call_index(0)]
+		pub fn set_forwarder(
+			origin: OriginFor<T>,
+			router_id: T::RouterId,
+			source_domain: Domain,
+			forwarding_contract: H160,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			RouterForwarding::<T>::insert(
+				&router_id,
+				ForwardInfo {
+					source_domain,
+					contract: forwarding_contract,
+				},
+			);
+
+			Self::deposit_event(Event::<T>::ForwarderSet {
+				router_id,
+				source_domain,
+				forwarding_contract,
+			});
+
+			Ok(())
+		}
+
+		/// Remove the forwarding info for the given router id.
+		///
+		/// Origin: Admin.
+		///
+		/// NOTE: Simple weight due to origin requirement.
+		#[pallet::weight(T::DbWeight::get().writes(1))]
+		#[pallet::call_index(1)]
+		pub fn remove_forwarder(origin: OriginFor<T>, router_id: T::RouterId) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			RouterForwarding::<T>::take(&router_id)
+				.map(|info| {
+					Self::deposit_event(Event::<T>::ForwarderRemoved {
+						router_id,
+						source_domain: info.source_domain,
+						forwarding_contract: info.contract,
+					});
+				})
+				.ok_or(Error::<T>::ForwardInfoNotFound.into())
+		}
+	}
+
+	impl<T: Config> MessageSender for Pallet<T> {
+		type Message = T::Message;
+		type Middleware = T::RouterId;
+		type Origin = DomainAddress;
+
+		fn send(
+			router_id: T::RouterId,
+			origin: DomainAddress,
+			message: T::Message,
+		) -> DispatchResult {
+			let msg = RouterForwarding::<T>::get(&router_id)
+				.map(|info| {
+					T::Message::try_wrap_forward(info.source_domain, info.contract, message.clone())
+				})
+				.unwrap_or_else(|| {
+					ensure!(!message.is_forwarded(), Error::<T>::ForwardInfoNotFound);
+					Ok(message)
+				})?;
+
+			T::MessageSender::send(router_id, origin, msg)
+		}
+	}
+
+	impl<T: Config> MessageReceiver for Pallet<T> {
+		type Message = T::Message;
+		type Middleware = T::RouterId;
+		type Origin = DomainAddress;
+
+		fn receive(
+			router_id: T::RouterId,
+			forwarding_domain_address: DomainAddress,
+			message: T::Message,
+		) -> DispatchResult {
+			// Message can be unwrapped iff it was forwarded
+			//
+			// NOTE: Contract address irrelevant here because it is only necessary for
+			// outbound forwarded messages
+			let (lp_message, domain_address) = match (
+				RouterForwarding::<T>::get(&router_id),
+				message.clone().unwrap_forwarded(),
+			) {
+				(Some(info), Some((source_domain, _contract, lp_message))) => {
+					ensure!(
+						info.source_domain == source_domain,
+						Error::<T>::SourceDomainMismatch
+					);
+
+					let domain_address = DomainAddress::Evm(
+						info.source_domain
+							.get_evm_chain_id()
+							.expect("Domain not Centrifuge; qed"),
+						info.contract,
+					);
+					Ok((lp_message, domain_address))
```

Comment:
Nit - this one is actually the `forwarding_domain_address`. Feel free to disregard tho, can live with this.

Reply:
No both are different:

* `forwarding_domain_address` is some address of the forwarding domain `Ethereum`, e.g. `DomainAddress::Evm(1, x)`
* `domain_address` here is the the domain address of the source domain and the forwarding contract, e.g. `DomainAddress::Evm(42220, info.contract)`